### ISSUE TITLE
Add kubernetes api schema validation to linter

### ISSUE THREAD

**Issue by wilreichert** [helm/helm#2575]
Kubernetes has a validation engine which can be used to ensure yaml conforms to the relevant API group before its deployed via the `--validate=true` flag. It would be exceedingly useful to leverage this functionality to identify template formatting issues. Since the linter is rules based it seems like a reasonable place to add this feature rather than adding a new command like `validate`.

**Comment by technosophos** [comment]
We looked at this before, and the consensus was that `helm lint` must not require an active network connection in order to stay backward compatible. However, it would be possible to do client-side validation by using the JSON Schema files.

**Comment by wilreichert** [comment]
I looked at using local schema files, but I'm not sure you can guarantee they will be present on a client system. They are cached on initial server access? Plus without a server available you are still faced with the dilemma of what schema version to use. Is there some default schema version available client side that I'm unaware of?

Another option would be to make this is own CLI parameter - 'validate' perhaps? I'm still inclined to think this is linting type functionality though.

**Comment by thomastaylor312** [comment]
Putting this back on the backlog milestone until someone takes it on

**Comment by jdotjdot** [comment]
What about providing a CLI option that would permit connections to a live cluster?  We've run into problems with Helm over and over where the Kubernetes YAML turned out to be incorrect (though valid YAML itself), and when it failed, the k8s error was swallowed by a Helm error, leading to red herrings.  This would have been prevented if we could clearly see from Kubernetes that the YAML wasn't a fully valid spec.

**Comment by kkimdev** [comment]
For kubectl, --validate is an option for apply and create commands.  How about adding --validate option to helm install & upgrade to be consistent with kubectl?  Then we can bypass the helm lint no networking problem too right?

**Comment by bacongobbler** [comment]
A feature flag that permits `helm lint` to validate against the openAPI schema would be great. I wonder how it'd CRDs would interact with that feature, since they are part of the chart but have not been applied to the cluster yet. If anyone has the time to implement it, I'd love to hear how that works!

As a side note, PRs that add new features are much more likely to be merged if the work was done against Helm 3. Right now we're focusing on bug and stability fixes for Helm 2. We simply don't have the time to focus on reviewing features for both Helm 2 and Helm 3, so we're trying to focus on the latter. :)

**Comment by bacongobbler** [comment]
#2585 was due to inactivity. If someone is interested in pushing this feature along the finish line, please write a PR against Helm 3. As per [the announcement on the mailing list](https://lists.cncf.io/g/cncf-helm/topic/last_v2_feature_release/34301238), Helm 2.15.0 is the last release where we will be merging new features for Helm 2. Thanks!

**Comment by lentzi90** [comment]
I would like to highlight that due to the missing validation functionality, things may actually fail in quite subtle ways and I suggest raising the priority of this issue because of this.

The "obvious" way things fail without validation, as already mentioned above, is that a release ends up failed because the manifests were invalid Kubernetes objects. You may get strange error messages from helm sometimes due to this and it is of course annoying.

What is worse though, is that the release may succeed just fine, without any warnings or other signs of problems, but the manifests may still be invalid and the API server will use default values instead of whatever you provided. This means that you may think that everything is configured in a certain way even though it is not. Very confusing and potentially dangerous.

How can this be? Let me show the example that thought me this:

A tired human happens to put the `strategy` field of a Deployment under `spec.template.spec` instead of `spec`, where it belongs. A simple mistake that is quite hard to spot. It looks something like this:

```yaml
...
spec:
  replicas: {{ .Values.replicaCount }}
  # This is where strategy should be
  template:
    spec:
      # This is not valid: strategy should be under spec, not spec.template.spec
      strategy:
        type: RollingUpdate
        rollingUpdate:
          maxUnavailable: 0
          maxSurge: 1
...
```

This is not valid, and if you try to `kubectl apply` the manifest, it will fail with a suggestion that you can set `--validation=false` if you want to ignore the problem. But if you use helm directly instead there is no error or warning. This is what actually ended up in Kubernetes:

```yaml
...
spec:
  replicas: 1
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
...
```

That is a correct manifest and the default if you don't specify anything else. The API server simply ignores "extra" fields and adds defaults as needed. This is quite dangerous, very confusing and frustrating to debug if you are lucky enough to spot the difference in the first place.

There are a couple of issues that I think should be counted as duplicates of this one, namely:

- #5836
- #5577 (note how this shows exactly how the API server ignores `completelyRandomName: my.conf`)

#6663 is also related but different. I'm not sure how the `helm template --validate` works, but a quick test indicates that it does **not** check that the manifest is valid, but rather that it is possible to apply it on the server. If I "validate" the example above I still get the invalid manifest as output, i.e. `strategy` is in the wrong place.

TL;DR: I think this issue deserves a higher priority. Alternatively, a new issue should be opened to adress validation during upgrade/install.

**Comment by adam-carbone** [comment]
should this work the same way as what people are describing? 

helm install --dry-run --debug . | kubectl apply -f -

**Comment by adam-carbone** [comment]
> should this work the same way as what people are describing?
> 
> helm install --dry-run --debug . | kubectl apply -f -

Nevermind, I see that it has a bunch of other things that don't allow the manifests to be processed

**Comment by bacongobbler** [comment]
There's `helm template --validate` available in Helm 3, which performs the same operation as what's described here. PRs welcome to port that over for `helm lint`.

**Comment by DanielCalvo** [comment]
Hey everyone! Googling "help api validation" brought me here.

Here's a curious thing we stumbled upon at my workplace this week: Cronjobs have a 52 character limit for their names.

```yaml
apiVersion: batch/v1beta1
kind: CronJob
metadata:
  name: did-you-know-that-k8s-limits-cronjob-names-to-52-characters
spec:
  schedule: "*/1 * * * *"
  jobTemplate:
    spec:
      template:
        spec:
          containers:
            - name: hello
              image: busybox
              imagePullPolicy: IfNotPresent
              args:
                - /bin/sh
                - -c
                - date; echo Hello from the Kubernetes cluster
          restartPolicy: OnFailure
```

Kubernetes will not accept this object:

```shell
kubectl apply -f cronjob.yaml 
The CronJob "did-you-know-that-k8s-limits-cronjob-names-to-52-characters" is invalid: metadata.name: Invalid value: "did-you-know-that-k8s-limits-cronjob-names-to-52-characters": must be no more than 52 characters
```

We can put this in a chart and then lint it, template it, and dry-run the install, but none of these catch the issue:

```shell
helm lint  ./testchart/
helm template --validate ./testchart/
helm install --dry-run --debug testchart ./testchart/
```

Somewhat off topic, but disappointingly, kubeval also doesn't see anything wrong with this:

```shell
cat cronjob.yaml | kubeval
PASS - stdin contains a valid CronJob (did-you-know-that-k8s-limits-cronjob-names-to-52-characters)
```

It is only when you do server side validation with kubectl that you can actually spot this:

```shell
kubectl apply --dry-run=server -f cronjob.yaml
The CronJob "did-you-know-that-k8s-limits-cronjob-names-to-52-characters" is invalid: metadata.name: Invalid value: "did-you-know-that-k8s-limits-cronjob-names-to-52-characters": must be no more than 52 characters
```

In the end, on top of `helm lint` on the linting step of our pipeline, we also added:
```shell
helm template --validate ./testchart/ | kubectl apply -f - --dry-run=server
```

So we can know that the Kubernetes API will accept our object definitions.

@bacongobbler maybe this is worth a new feature request? I'm not sure if I'm missing something here. It would be nice to be able to do this type of server side validation (--dry-run=server) with helm.

Tested with helm 3.5 and k3s 1.20

**Comment by lentzi90** [comment]
@DanielCalvo see also my [comment above](https://github.com/helm/helm/issues/2575#issuecomment-547373580). This behavior can actually *hide* errors since helm basically disables validation like if you run `kubectl apply --validation=false`. Of course if it is a critical error like the CronJob name length, it won't be accepted anyway, but if it is something like you specified a field in the wrong place it will just silently ignore it. :scream:

**Comment by bacongobbler** [comment]
Feel free to propose PRs to improve Helm's validation logic.
